### ISSUE TITLE
[BUG]: Memory of GPU has not been reduced when I have some network before main model.

### ISSUE THREAD

**Issue by yue-gang** [hpcaitech/ColossalAI#3642]
### 🐛 Describe the bug

When I have some network before main model, my memory of gpu has not reduced. As follows,

with torch.no_grad():
         z = autoencoder.encode(img)

engine.zero_grad()
output = engine(z)
loss = engine.criterion(output)

engine.backward(loss.mean())
engine.step()
lr_scheduler.step()

But, While I replace 'z = autoencoder.encode(img)' with ‘z = torch.randn(shape)', the memory will meet expectations. (less than ddp)

**Comment by JThh** [comment]
By 'reduced', which settings are you comparing against?

**Comment by yue-gang** [comment]
@JThh 
GPU: A100 80GB x 8
Config: tensor parallel 2d/3d, just like this,

colo_config = Config({
    "BATCH_SIZE" : 24,
    "NUM_MICRO_BATCHES": 24,
    "NUM_EPOCHS" : 2,
    "fp16" : {
        "mode" : AMP_TYPE.NAIVE
    },
    'parallel': dict(
        pipeline=1,
        tensor=dict(size=8, mode='3d'),
    )
})

**Comment by JThh** [comment]
May I further seek to clarify if you are comparing tensor2d to tensor3d, or simply tp to no parallelism?

**Comment by yue-gang** [comment]
> May I further seek to clarify if you are comparing tensor2d to tensor3d, or simply tp to no parallelism?

simply tp to no parallelism

**Comment by JThh** [comment]
Hi, you might want to [profile](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html) your autoencoder model memory usage.

**Comment by yue-gang** [comment]
> profile

With this code added, the memory has become normal.
> autoencoder, _, _, _ = colossalai.initialize(autoencoder, train_state.optimizer)

But I don't understand why.
